optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/NearbyDistanceMatrix.java

```diff
@@ -20,20 +20,29 @@
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
+import java.util.function.Function;
+import java.util.function.ToIntFunction;
 
 public final class NearbyDistanceMatrix {
 
     private final NearbyDistanceMeter nearbyDistanceMeter;
     private final Map<Object, Object[]> originToDestinationsMap;
+    private final Function<Object, Iterator<Object>> destinationIteratorProvider;
+    private final ToIntFunction<Object> destinationSizeFunction;
```

Comment:
That's a lot of `Object` there. Any chance to make the class generic?
Even if we can't use the generics on the outside, I think it still helps on the inside, makes the class better self-documenting I think.